server/src/test/java/io/druid/server/namespace/cache/TestNamespaceExtractionCacheManagerExecutors.java

```diff
@@ -0,0 +1,362 @@
+      Assert.assertTrue(
+          String.format(
+              "Didn't wait long enough between runs. Expected more than %d was %d",
+              minEnd - start,
+              end - start
+          ), minEnd < end
+      );
+    }
+    finally {
+      lifecycle.stop();
+    }
+    totalRunCount = ranCount.get();
+    Thread.sleep(50);
+    Assert.assertEquals(totalRunCount, ranCount.get(), 1);
+  }
+
+
+  @Test(timeout = 10_000)
+  public void testConcurrentDelete() throws ExecutionException, InterruptedException
+  {
+    final int threads = 5;
+    ListeningExecutorService executorService = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(threads));
+    final CountDownLatch latch = new CountDownLatch(threads);
+    Collection<ListenableFuture<?>> futures = new ArrayList<>();
+    for(int i = 0; i < threads; ++i){
+      final int loopNum = i;
+      ListenableFuture<?> future = executorService.submit(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              try {
+                latch.countDown();
+                latch.await();
+                for(int j = 0; j < 10; ++j) {
+                  testDelete(String.format("ns-%d", loopNum));
+                }
+              }
+              catch (InterruptedException e) {
+                throw Throwables.propagate(e);
+              }
+            }
+          }
+      );
+    }
+    Futures.allAsList(futures).get();
+    executorService.shutdown();
+  }
+
+  @Test(timeout = 500)
+  public void testDelete()
+      throws NoSuchFieldException, IllegalAccessException, InterruptedException, ExecutionException
+  {
+    try{
+      testDelete("ns");
+    }finally{
+      lifecycle.stop();
+    }
+  }
+
+  public void testDelete(final String ns)
+      throws InterruptedException
+  {
+    final CountDownLatch latch = new CountDownLatch(5);
+    final AtomicLong runs = new AtomicLong(0);
+    long prior = 0;
+    final URIExtractionNamespace namespace = new URIExtractionNamespace(
+        ns,
+        tmpFile.toURI(),
+        new URIExtractionNamespace.ObjectMapperFlatDataParser(
+            URIExtractionNamespaceTest.registerTypes(new ObjectMapper())
+        ),
+        1l
+    );
+    final String cacheId = UUID.randomUUID().toString();
+    ListenableFuture<?> future =
+        manager.schedule(
+            namespace, factory, new Runnable()
+            {
+              @Override
+              public void run()
+              {
+                manager.getPostRunnable(namespace, factory, cacheId).run();
+                runs.incrementAndGet();
+                latch.countDown();
+              }
+            },
+            cacheId
+        );
+    latch.await();
+    Assert.assertFalse(future.isCancelled());
+    Assert.assertFalse(future.isDone());
+    Assert.assertTrue(fnCache.containsKey(ns));
+    prior = runs.get();
+    Thread.sleep(50);
+    Assert.assertTrue(runs.get() > prior);
```

Comment:
I modified this a bit to hopefully change the way the tests are done and squash the problem, which I think was a test problem rather than an impl problem.